### PR TITLE
Simplify kotlin code and improve code readability

### PR DESCRIPTION
**Pull request by curtishd**
For chapter array_and_linkedlist, backtracking and computational_complexity, Adding missing comments, fix indentation and more.

### CODE REVIEW

File: codes/kotlin/chapter_backtracking/n_queens.kt [hunk 1]
```diff
@@ -51,9 +51,9 @@ fun backtrack(
 /* 求解 n 皇后 */
 fun nQueens(n: Int): List<List<List<String>>?> {
     // 初始化 n*n 大小的棋盘，其中 'Q' 代表皇后，'#' 代表空位
-    val state: MutableList<MutableList<String>> = ArrayList()
+    val state = ArrayList<MutableList<String>>()
```
Comment (krahets) [inline codes/kotlin/chapter_backtracking/n_queens.kt:54]: I'm confused about the `val state = ArrayList<MutableList<String>>()`, but `val copyState = ArrayList<List<String>>()`. Can you unify them?

File: codes/kotlin/chapter_backtracking/preorder_traversal_ii_compact.kt [hunk 2]
```diff
@@ -10,7 +10,7 @@ import utils.TreeNode
 import utils.printTree
 
 var path: MutableList<TreeNode>? = null
-var res: MutableList<List<TreeNode>>? = null
+var res: ArrayList<List<TreeNode>>? = null
```
Comment (krahets) [inline codes/kotlin/chapter_backtracking/preorder_traversal_ii_compact.kt:13]: I am confused with the difference among `List` `ArrayList` `MutableList`. Why do you use different types for the different variables.

Comment (curtishd) [reply codes/kotlin/chapter_backtracking/preorder_traversal_ii_compact.kt:13]: Sure

**Comment by curtishd** [comment]
I admin that missing up List, ArrayList and MutableList can be confusing, For the reason of following the principle of minimum variability. I have try my best to replace all ArrayList with Mutablelist, however, Mutablelist could not simply replace ArrayList in some situation for lacking of constructor.

**Comment by curtishd** [comment]
I admin that mixing up List, ArrayList and MutableList can be confusing, For the reason of following the principle of minimum variability. I have try my best to replace all ArrayList with Mutablelist, however, Mutablelist could not simply replace ArrayList in some situation for lacking of constructor.

File: codes/kotlin/chapter_array_and_linkedlist/list.kt [hunk 3]
```diff
@@ -8,6 +8,7 @@ package chapter_array_and_linkedlist
 
 /* Driver Code */
 fun main() {
+    /* 初始化列表 */
     // 可变集合
     val numbers = mutableListOf(1, 3, 2, 5, 4)
     val nums = ArrayList<Int>(numbers)
```
Comment (krahets) [inline codes/kotlin/chapter_array_and_linkedlist/list.kt:14]: Not consistent. Avoid using ArrayList for all the kotlin code

```
nums = ArrayList<Int>
nums1 = mutableListOf
```

File: codes/kotlin/chapter_backtracking/n_queens.kt [hunk 4]
```diff
@@ -10,15 +10,15 @@ package chapter_backtracking.n_queens
...
     cols: BooleanArray,
     diags1: BooleanArray,
     diags2: BooleanArray
 ) {
     // 当放置完所有行时，记录解
     if (row == n) {
-        val copyState: MutableList<List<String>> = ArrayList()
+        val copyState = mutableListOf<MutableList<String>>()
         for (sRow in state) {
             copyState.add(ArrayList(sRow))
```
Comment (krahets) [inline codes/kotlin/chapter_backtracking/n_queens.kt:23]: Avoid using ArrayList. Here is an exmaple:

```kt
if (row == n) {
    val copyState = mutableListOf<MutableList<String>>()
    for (sRow in state) {
        // Use toMutableList() to create a new MutableList that is a copy of sRow
        copyState.add(sRow.toMutableList())
    }
}
```

File: codes/kotlin/chapter_backtracking/permutations_ii.kt [hunk 5]
```diff
@@ -11,15 +11,15 @@ fun backtrack(
     state: MutableList<Int>,
     choices: IntArray,
     selected: BooleanArray,
-    res: MutableList<MutableList<Int>?>
+    res: MutableList<List<Int>?>
 ) {
     // 当状态长度等于元素数量时，记录解
     if (state.size == choices.size) {
         res.add(ArrayList(state))
```
Comment (krahets) [inline codes/kotlin/chapter_backtracking/permutations_ii.kt:18]: Same as above

Comment (krahets) [inline codes/kotlin/chapter_backtracking/permutations_ii.kt:18]: Same as above

File: codes/kotlin/chapter_backtracking/preorder_traversal_iii_template.kt [hunk 6]
```diff
@@ -17,7 +17,7 @@ fun isSolution(state: List<TreeNode?>): Boolean {
 
 /* 记录解 */
 fun recordSolution(state: MutableList<TreeNode?>?, res: MutableList<List<TreeNode?>?>) {
-    res.add(state?.let { ArrayList(it) })
+    res.add(ArrayList(state!!))
```
Comment (krahets) [inline codes/kotlin/chapter_backtracking/preorder_traversal_iii_template.kt:20]: Same as above

File: codes/kotlin/chapter_backtracking/n_queens.kt (hunk 4 above)
Comment (krahets) [inline codes/kotlin/chapter_backtracking/n_queens.kt:23]: copyState.add(sRow.toMutableList())

**Review by krahets — APPROVED** [review]
Thanks for the PR!

**Pull request merged**